<?xml version="1.0"?>

<robot name="sixDOF">
  
  <!-- *************************************************** -->
  <!-- ******************** Kinematics ******************* -->
  <!-- *************************************************** -->

  <link name="world"/>

  <joint name="theta1" type="continuous">
    <parent link="world"/>
    <child  link="link1"/>
    <origin xyz="0 0 0.6" rpy="0 0 0"/>
    <axis   xyz="0 0 1"/>
    <dynamics damping="0.5"/>
  </joint>

  <link name="link1"/>

  <joint name="theta2" type="continuous">
    <parent link="link1"/>
    <child  link="link2"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis   xyz="1 0 0"/>
    <dynamics damping="0.5"/>
  </joint>

  <link name="link2"/>

  <joint name="attach-shoulder" type="fixed">
    <parent link="link2"/>
    <child  link="shoulder"/>
  </joint>

  <link name="shoulder"/>

  <joint name="theta3" type="continuous">
    <parent link="link2"/>
    <child  link="link3"/>
    <origin xyz="0 0.4 0" rpy="0 0 0"/>
    <axis   xyz="1 0 0 "/>
    <dynamics damping="0.1"/>
  </joint>

  <link name="link3"/>

  <joint name="attach-elbow" type="fixed">
    <parent link="link3"/>
    <child  link="elbow"/>
  </joint>

  <link name="elbow"/>

  <joint name="theta4" type="continuous">
    <parent link="link3"/>
    <child  link="link4"/>
    <origin xyz="0 0.4 0" rpy="0 0 0"/>
    <axis   xyz="0 1 0"/>
    <dynamics damping="0.02"/>
  </joint>

  <link name="link4"/>

  <joint name="theta5" type="continuous">
    <parent link="link4"/>
    <child  link="link5"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis   xyz="0 0 1"/>
    <dynamics damping="0.02"/>
  </joint>

  <link name="link5"/>

  <joint name="theta6" type="continuous">
    <parent link="link5"/>
    <child  link="link6"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis   xyz="1 0 0"/>
    <dynamics damping="0.02"/>
  </joint>

  <link name="link6"/>

  <joint name="attach-wrist" type="fixed">
    <parent link="link6"/>
    <child  link="wrist"/>
  </joint>

  <link name="wrist"/>

  <joint name="final" type="fixed">
    <parent link="link6"/>
    <child  link="tip"/>
    <origin xyz="0 0.15 0" rpy="0 0 0"/>
  </joint>

  <link name="tip"/>


  <!-- *************************************************** -->
  <!-- ******************* RVIZ Colors ******************* -->
  <!-- *************************************************** -->

  <material name="white">    <color rgba="1.00 1.00 1.00 1"/>  </material>
  <material name="gray">     <color rgba="0.60 0.60 0.60 1"/>  </material>
  <material name="table">    <color rgba="0.85 0.77 0.77 1"/>  </material>
  <material name="black">    <color rgba="0.00 0.00 0.00 1"/>  </material>

  <material name="red">      <color rgba="1.00 0.00 0.00 1"/>  </material>
  <material name="green">    <color rgba="0.00 1.00 0.00 1"/>  </material>
  <material name="blue">     <color rgba="0.00 0.00 1.00 1"/>  </material>

  <material name="cyan">     <color rgba="0.00 1.00 1.00 1"/>  </material>
  <material name="magenta">  <color rgba="1.00 0.00 1.00 1"/>  </material>
  <material name="yellow">   <color rgba="1.00 1.00 0.00 1"/>  </material>

  <material name="orange">   <color rgba="1.00 0.65 0.00 1"/>  </material>


  <!-- *************************************************** -->
  <!-- ****** Building Blocks (Visual AND Inertial) ****** -->
  <!-- *************************************************** -->

  <!-- *** World-Fixed: Mount 1, Motor 1 *** -->

  <joint name="attach-mount1" type="fixed">
    <parent link="world"/>
    <child  link="mount1"/>
  </joint>
  <link name="mount1">
    <visual>
      <origin xyz="0 0 0.1" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.02"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0 0.1" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.02"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0.1" rpy="0 0 0"/>
      <mass value="0.2"/>
      <inertia
        ixx="0.000687" ixy="0.0" ixz="0.0"
        iyy="0.000687" iyz="0.0"
        izz="0.000040"/>
    </inertial>
  </link>

  <joint name="attach-motor1" type="fixed">
    <parent link="world"/>
    <child  link="motor1"/>
  </joint>
  <link name="motor1">
    <visual>
      <origin xyz="0 0 0.3" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.05"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <collision>
      <origin xyz="0 0 0.3" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.05"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0.3" rpy="0 0 0"/>
      <mass value="2.0"/>
      <inertia
        ixx="0.007917" ixy="0.0" ixz="0.0"
        iyy="0.007917" iyz="0.0"
        izz="0.002500"/>
    </inertial>
  </link>

  <!-- *** Motor 1 Output: Mount 2, Motor 2 *** -->

  <joint name="attach-mount2" type="fixed">
    <parent link="link1"/>
    <child  link="mount2"/>
  </joint>
  <link name="mount2">
    <visual>
      <origin xyz="0 0 -0.1" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.02"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0 -0.1" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.02"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 -0.1" rpy="0 0 0"/>
      <mass value="0.2"/>
      <inertia
        ixx="0.000687" ixy="0.0" ixz="0.0"
        iyy="0.000687" iyz="0.0"
        izz="0.000040"/>
    </inertial>
  </link>

  <joint name="attach-motor2" type="fixed">
    <parent link="link1"/>
    <child  link="motor2"/>
  </joint>
  <link name="motor2">
    <visual>
      <origin xyz="0 0 0" rpy="0 1.5708 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.05"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 1.5708 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.05"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 1.5708 0"/>
      <mass value="2.0"/>
      <inertia
        ixx="0.007917" ixy="0.0" ixz="0.0"
        iyy="0.007917" iyz="0.0"
        izz="0.002500"/>
    </inertial>
  </link>

  <!-- *** Motor 2 Output: Mount 3, Motor 3 *** -->

  <joint name="attach-mount3" type="fixed">
    <parent link="link2"/>
    <child  link="mount3"/>
  </joint>
  <link name="mount3">
    <visual>
      <origin xyz="0 0.2 0" rpy="1.5708 0 0"/>
      <geometry>
	<cylinder length="0.4" radius="0.02"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0.2 0" rpy="1.5708 0 0"/>
      <geometry>
	<cylinder length="0.4" radius="0.02"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0.2 0" rpy="1.5708 0 0"/>
      <mass value="0.4"/>
      <inertia
        ixx="0.005373" ixy="0.0" ixz="0.0"
        iyy="0.005373" iyz="0.0"
        izz="0.000080"/>
    </inertial>
  </link>

  <joint name="attach-motor3" type="fixed">
    <parent link="link2"/>
    <child  link="motor3"/>
  </joint>
  <link name="motor3">
    <visual>
      <origin xyz="0 0.4 0" rpy="0 1.5708 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.05"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <collision>
      <origin xyz="0 0.4 0" rpy="0 1.5708 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.05"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0.4 0" rpy="0 1.5708 0"/>
      <mass value="2.0"/>
      <inertia
        ixx="0.007917" ixy="0.0" ixz="0.0"
        iyy="0.007917" iyz="0.0"
        izz="0.002500"/>
    </inertial>
  </link>

  <!-- *** Motor 3 Output: Mount 4, Motor 4 *** -->

  <joint name="attach-mount4" type="fixed">
    <parent link="link3"/>
    <child  link="mount4"/>
  </joint>
  <link name="mount4">
    <visual>
      <origin xyz="0 0.1 0" rpy="1.5708 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.02"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0.1 0" rpy="1.5708 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.02"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0.1 0" rpy="1.5708 0 0"/>
      <mass value="0.2"/>
      <inertia
        ixx="0.000687" ixy="0.0" ixz="0.0"
        iyy="0.000687" iyz="0.0"
        izz="0.000040"/>
    </inertial>
  </link>

  <joint name="attach-motor4" type="fixed">
    <parent link="link3"/>
    <child  link="motor4"/>
  </joint>
  <link name="motor4">
    <visual>
      <origin xyz="0 0.2 0" rpy="1.5708 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.025"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <collision>
      <origin xyz="0 0.2 0" rpy="1.5708 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.025"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0.2 0" rpy="1.5708 0 0"/>
      <mass value="1.0"/>
      <inertia
        ixx="0.000990" ixy="0.0" ixz="0.0"
        iyy="0.000990" iyz="0.0"
        izz="0.000313"/>
    </inertial>
  </link>

  <!-- *** Motor 4 Output: Mount 5, Motor 5 *** -->

  <joint name="attach-mount5" type="fixed">
    <parent link="link4"/>
    <child  link="mount5"/>
  </joint>
  <link name="mount5">
    <visual>
      <origin xyz="0 -0.1 0" rpy="1.5708 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 -0.1 0" rpy="1.5708 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.01"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 -0.1 0" rpy="1.5708 0 0"/>
      <mass value="0.1"/>
      <inertia
        ixx="0.000336" ixy="0.0" ixz="0.0"
        iyy="0.000336" iyz="0.0"
        izz="0.000005"/>
    </inertial>
  </link>

  <joint name="attach-motor5" type="fixed">
    <parent link="link4"/>
    <child  link="motor5"/>
  </joint>
  <link name="motor5">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.025"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.025"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="1.0"/>
      <inertia
        ixx="0.000990" ixy="0.0" ixz="0.0"
        iyy="0.000990" iyz="0.0"
        izz="0.000313"/>
    </inertial>
  </link>

  <!-- *** Motor 5 Output: Motor 6 *** -->

  <joint name="attach-motor6" type="fixed">
    <parent link="link5"/>
    <child  link="motor6"/>
  </joint>
  <link name="motor6">
    <visual>
      <origin xyz="0 0 0" rpy="0 1.5708 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.025"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 1.5708 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.025"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 1.5708 0"/>
      <mass value="1.0"/>
      <inertia
        ixx="0.000990" ixy="0.0" ixz="0.0"
        iyy="0.000990" iyz="0.0"
        izz="0.000313"/>
    </inertial>
  </link>

  <!-- *** Motor 6 Output: Base, Palm, Fingers *** -->

  <joint name="attach-handbase" type="fixed">
    <parent link="link6"/>
    <child  link="handbase"/>
  </joint>
  <link name="handbase">
    <visual>
      <origin xyz="0 0.05 0" rpy="1.5708 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0.05 0" rpy="1.5708 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.01"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0.05 0" rpy="1.5708 0 0"/>
      <mass value="0.05"/>
      <inertia
        ixx="0.000043" ixy="0.0" ixz="0.0"
        iyy="0.000043" iyz="0.0"
        izz="0.000003"/>
    </inertial>
  </link>

  <joint name="attach-palm" type="fixed">
    <parent link="link6"/>
    <child  link="palm"/>
  </joint>
  <link name="palm">
    <visual>
      <origin xyz="0 0.1 0" rpy="0 1.5708 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0.1 0" rpy="0 1.5708 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.01"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0.1 0" rpy="0 1.5708 0"/>
      <mass value="0.05"/>
      <inertia
        ixx="0.000043" ixy="0.0" ixz="0.0"
        iyy="0.000043" iyz="0.0"
        izz="0.000003"/>
    </inertial>
  </link>

  <joint name="attach-leftfinger" type="fixed">
    <parent link="link6"/>
    <child  link="leftfinger"/>
  </joint>
  <link name="leftfinger">
    <visual>
      <origin xyz="-0.05 0.14 0" rpy="1.5708 0 0"/>
      <geometry>
	<cylinder length="0.08" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="-0.05 0.14 0" rpy="1.5708 0 0"/>
      <mass value="0.05"/>
      <inertia
        ixx="0.000022" ixy="0.0" ixz="0.0"
        iyy="0.000022" iyz="0.0"
        izz="0.000002"/>
    </inertial>
  </link>

  <joint name="attach-rightfinger" type="fixed">
    <parent link="link6"/>
    <child  link="rightfinger"/>
  </joint>
  <link name="rightfinger">
    <visual>
      <origin xyz="0.05 0.14 0" rpy="1.5708 0 0"/>
      <geometry>
	<cylinder length="0.08" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="0.05 0.14 0" rpy="1.5708 0 0"/>
      <mass value="0.05"/>
      <inertia
        ixx="0.000022" ixy="0.0" ixz="0.0"
        iyy="0.000022" iyz="0.0"
        izz="0.000002"/>
    </inertial>
  </link>

  <!-- *** Cube: Six sides *** -->

  <joint name="attach-cube" type="fixed">
    <parent link="tip"/>
    <child  link="cube"/>
  </joint>
  <link name="cube">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
	<box size="0.10 0.10 0.10"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="1.0"/>
      <inertia
        ixx="0.000120" ixy="0.0" ixz="0.0"
        iyy="0.000120" iyz="0.0"
        izz="0.000120"/>
    </inertial>
  </link>

  <joint name="attach-cube-xp" type="fixed">
    <parent link="cube"/>
    <child  link="cube-xp"/>
  </joint>
  <link name="cube-xp">
    <visual>
      <origin         xyz=" 0.04   0    0"    rpy="0 0 0"/>
      <geometry> <box size="0.0202 0.09 0.09"/> </geometry>
      <material name="red"/>
    </visual>
  </link>

  <joint name="attach-cube-xm" type="fixed">
    <parent link="cube"/>
    <child  link="cube-xm"/>
  </joint>
  <link name="cube-xm">
    <visual>
      <origin         xyz="-0.04   0    0"    rpy="0 0 0"/>
      <geometry> <box size="0.0202 0.09 0.09"/> </geometry>
      <material name="cyan"/>
    </visual>
  </link>

  <joint name="attach-cube-yp" type="fixed">
    <parent link="cube"/>
    <child  link="cube-yp"/>
  </joint>
  <link name="cube-yp">
    <visual>
      <origin         xyz=" 0    0.04   0"    rpy="0 0 0"/>
      <geometry> <box size="0.09 0.0202 0.09"/> </geometry>
      <material name="green"/>
    </visual>
  </link>

  <joint name="attach-cube-ym" type="fixed">
    <parent link="cube"/>
    <child  link="cube-ym"/>
  </joint>
  <link name="cube-ym">
    <visual>
      <origin         xyz=" 0   -0.04   0"    rpy="0 0 0"/>
      <geometry> <box size="0.09 0.0202 0.09"/> </geometry>
      <material name="magenta"/>
    </visual>
  </link>

  <joint name="attach-cube-zp" type="fixed">
    <parent link="cube"/>
    <child  link="cube-zp"/>
  </joint>
  <link name="cube-zp">
    <visual>
      <origin         xyz=" 0    0    0.04"   rpy="0 0 0"/>
      <geometry> <box size="0.09 0.09 0.0202"/> </geometry>
      <material name="blue"/>
    </visual>
  </link>

  <joint name="attach-cube-zm" type="fixed">
    <parent link="cube"/>
    <child  link="cube-zm"/>
  </joint>
  <link name="cube-zm">
    <visual>
      <origin         xyz=" 0    0   -0.04"   rpy="0 0 0"/>
      <geometry> <box size="0.09 0.09 0.0202"/> </geometry>
      <material name="yellow"/>
    </visual>
  </link>


  <!-- *************************************************** -->
  <!-- ******************* Gazebo Colors ***************** -->
  <!-- *************************************************** -->

  <gazebo reference="motor1"> <material>Gazebo/Gray</material>  </gazebo>
  <gazebo reference="motor2"> <material>Gazebo/Gray</material>  </gazebo>
  <gazebo reference="motor3"> <material>Gazebo/Gray</material>  </gazebo>
  <gazebo reference="motor4"> <material>Gazebo/Gray</material>  </gazebo>
  <gazebo reference="motor5"> <material>Gazebo/Gray</material>  </gazebo>
  <gazebo reference="motor6"> <material>Gazebo/Gray</material>  </gazebo>

  <gazebo reference="mount1">   <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="mount2">   <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="mount3">   <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="mount4">   <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="mount5">   <material>Gazebo/Black</material> </gazebo>

  <gazebo reference="handbase">    <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="palm">        <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="leftfinger">  <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="rightfinger"> <material>Gazebo/Black</material> </gazebo>

  <gazebo reference="cube-xp"> <material>Gazebo/Red</material>     </gazebo>
  <gazebo reference="cube-xm"> <material>Gazebo/Cyan</material>    </gazebo>
  <gazebo reference="cube-yp"> <material>Gazebo/Green</material>   </gazebo>
  <gazebo reference="cube-ym"> <material>Gazebo/Magenta</material> </gazebo>
  <gazebo reference="cube-zp"> <material>Gazebo/Blue</material>    </gazebo>
  <gazebo reference="cube-zm"> <material>Gazebo/Yellow</material>  </gazebo>


</robot>
